### PR TITLE
SQL Server e2e

### PR DESCRIPTION
**Pull request by Vipinofficial11**

### CODE REVIEW

File: mssql-plugin/src/e2e-test/features/mssql/mssql sink/DesignTimeWithMacros.feature [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+##
+## Unless required by applicable law or agreed to in writing, software
+## distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+## WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+## License for the specific language governing permissions and limitations under
+# the License..
+
+@Mysql
+Feature: MSSQL Server Sink - Design time scenarios (macro)
+
```
Comment (Aryan-Verma) [inline mssql-plugin/src/e2e-test/features/mssql/mssql sink/DesignTimeWithMacros.feature]: Design Time tests with macros where all the macros are filled and we are able to validate the plugin properties are redundant. We anyways check these in RuntimeWithMacros tests. So we can remove these tests.

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java [hunk 2]
```diff
@@ -16,76 +16,173 @@
...
+  @After(order = 1, value = "@MSSQL_SINK_TEST")
+  public static void dropTables() throws SQLException, ClassNotFoundException {
+    MssqlClient.deleteTables(PluginPropertyUtils.pluginProp("schema"),
+                             new String[]{PluginPropertyUtils.pluginProp("sourceTable"),
+                               PluginPropertyUtils.pluginProp("targetTable")});
+  }
+
+  @Before(order = 1, value = "@BQ_SINK")
+  public static void setTempTargetBQTable() {
+    bqTargetTable = "TestSN_table" + RandomStringUtils.randomAlphanumeric(10);
```
Comment (Aryan-Verma) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java]: Can we use UUID instead?

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java [hunk 3]
```diff
@@ -16,76 +16,173 @@
...
+  }
 
-    @After(order = 1, value = "@MSSQL_SINK_TEST")
-    public static void dropTables() throws SQLException, ClassNotFoundException {
-        MssqlClient.deleteTables(PluginPropertyUtils.pluginProp("schema"),
-                new String[]{PluginPropertyUtils.pluginProp("sourceTable"),
-                        PluginPropertyUtils.pluginProp("targetTable")});
+  private static void createSourceBQTableWithQueries(String bqCreateTableQueryFile, String bqInsertDataQueryFile)
+    throws IOException, InterruptedException {
+    String bqSourceTable = "E2E_SOURCE_" + UUID.randomUUID().toString().substring(0, 5).replaceAll("-",
```
Comment (Aryan-Verma) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java]: 5 length UUID substring won't be sufficient for the uniqueness condition I believe

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/mssql/BQValidation.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+          case Types.DATE:
+            Date dateSource = rsSource.getDate(currentColumnCount);
+            Date dateTarget = Date.valueOf(bigQueryData.get(jsonObjectIdx).get(columnName).getAsString());
+            Assert.assertEquals("Different values found for column : %s", dateSource, dateTarget);
+            break;
+          case Types.TIMESTAMP:
+            Timestamp timestampSource = rsSource.getTimestamp(currentColumnCount);
+            String targetTimestampString = bigQueryData.get(jsonObjectIdx).get(columnName).getAsString();
+            SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
+            dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
```
Comment (Aryan-Verma) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/mssql/BQValidation.java:224]: Please use DateTimeFormatter instead of SimpleDateFormat

Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/java/io.cdap.plugin/mssql/BQValidation.java:224]: For the case timestamp, It is required to remove the 'T' and 'Z' from the value for now. Once the bug is resolved and timestamp is mapped to DateTime, then we can use DateTimeFormatter.
Note - This validation logic is updated once the bug gets resolved.

File: mssql-plugin/src/e2e-test/features/mssql/mssql sink/DesignTimeWithMacros.feature (hunk 1 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql sink/DesignTimeWithMacros.feature]: Resolved

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java (hunk 2 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java]: Resolved

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java (hunk 3 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java]: Resolved

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/MssqlClient.java [hunk 5]
```diff
@@ -181,13 +185,11 @@ public static void createTargetDateTimeTable(String targetTable, String schema)
         }
     }
 
-    public static void deleteTables(String schema, String[] tables)
-            throws SQLException, ClassNotFoundException {
+    public static void deleteTable(String schema, String table)
+      throws SQLException, ClassNotFoundException {
         try (Connection connect = getMssqlConnection(); Statement statement = connect.createStatement()) {
-            for (String table : tables) {
                 String dropTableQuery = "DROP TABLE " + schema + "." + table;
```
Comment (saimukkamala) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/MssqlClient.java]: can we fix the indentation?

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java [hunk 6]
```diff
@@ -16,76 +16,292 @@
...
-    public static void createTables() throws SQLException, ClassNotFoundException {
-        MssqlClient.createSourceTable(PluginPropertyUtils.pluginProp("sourceTable"),
-                PluginPropertyUtils.pluginProp("schema"));
-        MssqlClient.createTargetTable(PluginPropertyUtils.pluginProp("targetTable"),
-                PluginPropertyUtils.pluginProp("schema"));
-    }
+  @Before(order = 2, value = "@MSSQL_AS_SOURCE")
+  public static void createSourceTables() throws SQLException, ClassNotFoundException {
+    MssqlClient.createSourceTable(PluginPropertyUtils.pluginProp("sourceTable"),
+                                  PluginPropertyUtils.pluginProp("schema"));
```
Comment (saimukkamala) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java]: nit: `sourceTable`, `targetTable`, `schema` strings have been across this file. It's a good idea to create constants for these strings.

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/mssql/BQValidation.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+      targetBigQueryResponse.add(json);
+    }
+
+    String getSourceQuery = "SELECT * FROM " + schema + "." + sourceTable;
+    try (Connection connect = MssqlClient.getMssqlConnection()) {
+      connect.setHoldability(ResultSet.HOLD_CURSORS_OVER_COMMIT);
+      Statement statement1 = connect.createStatement(ResultSet.TYPE_SCROLL_SENSITIVE, ResultSet.CONCUR_UPDATABLE,
+                                                     ResultSet.HOLD_CURSORS_OVER_COMMIT);
+
+      ResultSet rsSource = statement1.executeQuery(getSourceQuery);
```
Comment (saimukkamala) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/mssql/BQValidation.java]: nit: This 6 lines of code is repetitive. we can move this to a function which takes query and response as inputs and returns a result set.

**Review by saimukkamala — APPROVED** [review]

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/MssqlClient.java (hunk 5 above)
Comment (bharatgulati) [reply mssql-plugin/src/e2e-test/java/io.cdap.plugin/MssqlClient.java]: Done.

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java (hunk 6 above)
Comment (bharatgulati) [reply mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java]: Done.

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/mssql/BQValidation.java (hunk 7 above)
Comment (bharatgulati) [reply mssql-plugin/src/e2e-test/java/io.cdap.plugin/mssql/BQValidation.java]: Done.

**Comment by bharatgulati** [comment]
A few tests are failing due to an intermittent chrome version issue, and we have reported a bug for this. (https://cdap.atlassian.net/browse/PLUGIN-1649)

**Comment by Vipinofficial11** [comment]
Merging the PR with intermittent failure as discussed in the call.

**Pull request merged**